assets/js/app.js

```diff
@@ -0,0 +1,114 @@
+    const tasksContainer = document.getElementById('tasks')
+    tasksContainer.innerHTML += newHTML
+}
+
+function clearForm() {
+    const tasksForm = document.getElementById('tasksForm')
+    tasksForm.reset()
+
+    const descriptionInput = document.getElementById('description')
+    descriptionInput.focus()
+}
+
+function saveSingleTaskToStorage(taskObject) {
+    const currentTasksInStorageJSON = localStorage.getItem('tasks')
+    const currentTasksInStorage = JSON.parse(currentTasksInStorageJSON)
+    currentTasksInStorage.push(taskObject)
+    localStorage.setItem('tasks', JSON.stringify(currentTasksInStorage))
+}
+
+function loadTasksFromLocalStorage() {
+    // load the tasks from the local storage,
+    // arrange them by index order and saves them back in the right order
+    const tasksJSON = localStorage.getItem('tasks')
+    if (tasksJSON) {
+        const tasks = JSON.parse(tasksJSON)
+        const validTasks = [] // new array for the tasks that still not expired
+        for (const task of tasks) {
+            //check if the task is not expired
+            if (!isTaskExpired(task)) {
+                task.index = validTasks.length //change the index to new updated index (mostly after deleting one of the notes)
+                const newHTML = generateHTML(task) // saves to newHTML the existing tasks with updated index
+                renderHTML(newHTML) // render the taskContainer with the new index
+                validTasks.push(task)
+            }
+        }
+        localStorage.setItem('tasks', JSON.stringify(validTasks))
+    }
+}
+
+function isTaskExpired(task) {
+    const now = new Date();
+    const taskDateTime = new Date(task.date + 'T' + task.time);
+    return taskDateTime <= now;
+}
+
+function deleteAllCurrentTasks() {
+    // Clear the tasks from the DOM
+    const tasksContainer = document.getElementById('tasks');
+    tasksContainer.innerHTML = '';
+}
+
+function initStorage() {
+    // set the local storage to [] if empty
+    const currentTasksInStorageJSON = localStorage.getItem('tasks')
+    if (!currentTasksInStorageJSON) {
+        localStorage.setItem('tasks', JSON.stringify([]))
+    }
+}
+
+function getNumberOfTasksInLocalStorage() {
+    return JSON.parse(localStorage.getItem('tasks')).length
+}
+
+function addTask(event) {
+    event.preventDefault()
+    const data = collectData()
+    const newHTML = generateHTML(data, true)
+    renderHTML(newHTML)
+    saveSingleTaskToStorage(data) // saves single task to storage
+    clearForm() //clears the form after adding task
+}
+
+function deleteTask(index) {
+    // delete a task, then reupload all existing tasks 
+    const tasks = JSON.parse(localStorage.getItem('tasks'))
+    const tasksArray = []
+    for (const task of tasks) {
+        if (task.index !== index) tasksArray.push(task)
+    }
+    localStorage.setItem('tasks', JSON.stringify(tasksArray))
+    deleteAllCurrentTasks()
+    loadTasksFromLocalStorage()
```

Comment:
calling this here causes the delete button to also trigger expired task deletion, which is weird UX wise: i click delete on a task, suddenly several tasks disappear